### PR TITLE
plugin/helpers: fix an unchecked issue when comparing two configs

### PR DESCRIPTION
**Pull request by isacikgoz**
#### Summary
This PR proposes a fix for comparing the server configuration with the required configuration provided from a plugin. Current implementation only checks if the configuration exists but does not check the configuration value, I added an additional check and a test case to cover it.

#### Ticket Link
See https://github.com/mattermost/mattermost-plugin-demo/pull/74

### CODE REVIEW

**Comment by hanzei** [comment]
@prapti I'm inclined to skip QA review on this PR and rely on the QA review of https://github.com/mattermost/mattermost-plugin-demo/pull/74 as this PR is hard to test isolated.

Let me know, if you have objections.

File: plugin/helpers_config.go [hunk 1]
```diff
@@ -28,5 +30,5 @@ func (p *HelpersImpl) CheckRequiredServerConfiguration(req *model.Config) (bool,
 		return false, nil
 	}
 
-	return true, nil
+	return reflect.DeepEqual(&mergedCfg, req), nil
```
Comment (lieut-data) [inline plugin/helpers_config.go]: @isacikgoz, this seems counterintuitive -- `mergedCfg.ToJson() != cfg.ToJson()` should have exactly the desired semantics.

Actually, I wonder if the issue is with the order of parameters passed to `utils.Merge` -- right now, we start with the requested config, and apply the current config, but I think we want the opposite: start with the current config, and apply the requested config. If I make that change, the new test passes without the `reflect.DeepEqual` check here.

However, this crucially assumes that the incoming requested config leaves `nil` all parameters that it doesn't care about. Can you expand the unit tests to define a current configuration for other values in the same/different configuration structs and verify they remain unchanged? I'm wondering too if we can define a test that requires more than just one setting as well.

Comment (isacikgoz) [reply plugin/helpers_config.go]: @lieut-data, you are right. If we change the order there is no need to check `reflect.DeepEqual`. I will commit that change. In addition to that I expanded the the test cases to cover the issues you've pointed.

**Comment by prapti** [comment]
> @prapti I'm inclined to skip QA review on this PR and rely on the QA review of [mattermost/mattermost-plugin-demo#74](https://github.com/mattermost/mattermost-plugin-demo/pull/74) as this PR is hard to test isolated.
@hanzei I ran a quick sanity test (minimal testing with setting EnableTeamCreation true in config.json and false in System Console UI) to agree to not doing a long QA review here, but I'm still seeing the same behavior as I mentioned in https://github.com/mattermost/mattermost-plugin-demo/pull/74 where the setting in config.json takes precedence over the UI setting.
Could this be what you were referring to as being hard to test isolated?
I did approve the demo plugin PR, but would love to test further as needed.

**Comment by hanzei** [comment]
@prapti The issue is, that the code changes in this PR has to be compiled into the demo plugin PR. There is no way of testing the changes in this PR without using a plugins like the demo plugin.

After this PR is merged, @isacikgoz can update the demo plugins PR to use the code changes from this PR.

And if there are still issues with within the demo plugin PR, he would have to submit yet another PR. I know that's unfortunate, but it's the "easiest" way to test these kind of changes.

**Comment by lieut-data** [comment]
@hanzei / @prapti, might we be able to setup a test of the demo plugin that just points at this commit for the `mattermost-server` dependency? Then we can verify before merging.

**Comment by hanzei** [comment]
> @hanzei / @prapti, might we be able to setup a test of the demo plugin that just points at this commit for the `mattermost-server` dependency? Then we can verify before merging.

Yes that would work. But it would have to point to @isacikgoz's fork of `mattermost-server`. Hence, there would have to be a final commit that changes it back to upstream.

**Comment by prapti** [comment]
Thank you for addressing my concern @hanzei and @lieut-data. It does seem like it's cumbersome to set it up for testing the PR in isolation, eps. if the PR changes need to be compiled to reflect them.
Once this PR is merged then, could you please let me know so I know to test the demo plugin for the initial issue? That would be great. Thank you :)

**Comment by hanzei** [comment]
@lieut-data Would you be fine merging this PR?

**Pull request merged**